### PR TITLE
FossID: add server url and scan id to scan results

### PR DESCRIPTION
**Pull request by yarosevych**
Add server url and scan id to scan results

### CODE REVIEW

**Review by MarcelBochtler — APPROVED** [review]

**Comment by MarcelBochtler** [comment]
LGTM, maybe somebody else also wants to review this, as I already did an internal review.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 1]
```diff
@@ -358,30 +366,31 @@ class FossId internal constructor(
...
 
-            scanCode
+            scanCode to scanId
         } else {
             logger.info { "Scan '${existingScan.code}' found for $url and revision $revision." }
 
             requireNotNull(existingScan.code) {
                 "The code for an existing scan must not be null."
-            }
+            } to existingScan.id.toString()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:392]: Having an extra line like `existingScan.code to existingScan.id.toString()` would probably be a bit clearer.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 2]
```diff
@@ -358,30 +366,31 @@ class FossId internal constructor(
         revision: String,
         projectCode: String,
         projectName: String
-    ): String {
+    ): Pair<String, String> {
         val existingScan = scans.recentScansForRepository(url, revision).findLatestPendingOrFinishedScan()
 
-        val scanCode = if (existingScan == null) {
+        val scanCodeAndIdPair = if (existingScan == null) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:375]: It's probably fine to drop "Pair" from the name.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 3]
```diff
@@ -162,7 +165,7 @@ class FossId internal constructor(
         DELTA
     }
 
-    private val secretKeys = listOf("serverUrl", "apiKey", "user")
+    private val secretKeys = listOf("apiKey", "user")
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:168]: This change seems unrelated to the topic, strictly speaking, or?

Comment (yarosevych) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:168]: @sschuberth in my opinion since we add serverUrl to scan results, it doesn't make sense to hide it in scan options anymore. Please let me know if you disagree.

Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:168]: Ok, makes sense.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 2 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: The name is still the same (in this commit). Did you intend to change it?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 1 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: Likewise, this comment was not addressed (in this commit).

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 4]
```diff
@@ -358,30 +366,31 @@ class FossId internal constructor(
...
             logger.info { "Scan '${existingScan.code}' found for $url and revision $revision." }
 
             requireNotNull(existingScan.code) {
                 "The code for an existing scan must not be null."
-            }
+            } to existingScan.id.toString()
         }
-        if (config.waitForResult) checkScan(scanCode)
-        return scanCode
+        if (config.waitForResult) checkScan(scanCodeAndIdPair.first)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt]: As you touch this anyway, let's also have an empty line before the `if`.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 5]
```diff
@@ -383,14 +386,14 @@ class FossId internal constructor(
             scanCode to scanId
         } else {
             logger.info { "Scan '${existingScan.code}' found for $url and revision $revision." }
-
-            requireNotNull(existingScan.code) {
+            val existingScanCode = requireNotNull(existingScan.code) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:389]: Please keep the empty line before this one.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 6]
```diff
@@ -383,14 +386,14 @@ class FossId internal constructor(
...
                 "The code for an existing scan must not be null."
-            } to existingScan.id.toString()
+            }
+            existingScanCode to existingScan.id.toString()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:392]: Add an empty line before this one to make the return value stand out.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 1 above)
Comment (yarosevych) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: @sschuberth wrong commit, I have moved the changes to the respective commit now.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 2 above)
Comment (yarosevych) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: @sschuberth I have moved the changes to the respective commit.

File: scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt [hunk 7]
```diff
@@ -801,7 +803,12 @@ class FossIdTest : WordSpec({
 
             val result = fossId.scan(createPackage(pkgId, vcsInfo))
 
-            val expectedAdditionalData = mapOf(SCAN_CODE_KEY to scanCode, SCAN_ID_KEY to "1")
+            val expectedAdditionalData =
+                mapOf(
+                    SCAN_CODE_KEY to scanCode,
+                    SCAN_ID_KEY to "1",
+                    SERVER_URL_KEY to "https://www.example.org/fossid"
+                )
```
Comment (sschuberth) [inline scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt:811]: Nit: Formatting should be like

```
val expectedAdditionalData = mapOf(
    SCAN_CODE_KEY to scanCode,
    SCAN_ID_KEY to "1",
    SERVER_URL_KEY to "https://www.example.org/fossid"
)
```

But I'm also ok with keeping as-is if it passes *detekt* checks.

**Review by sschuberth — APPROVED** [review]

File: scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt (hunk 7 above)
Comment (sschuberth) [reply scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt:811]: You anyway have a [*detekt* issue about the test growing too large](https://github.com/oss-review-toolkit/ort/actions/runs/3966803517/jobs/6798069203#step:4:120), so you might want to address my above comment, too.

For the test size, I'd be fine with suppressing the hint by adding `@Suppress("LargeClass")` to the line before the test in the commit that starts to exceed the size.

**Comment by sschuberth** [comment]
Regarding the bogus *commitlint* failure on the "FossID" scope, I hope that https://github.com/oss-review-toolkit/ort/pull/6370 addresses it.

File: scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt (hunk 7 above)
Comment (yarosevych) [reply scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt]: I have addressed your first comment. As for the LargeClass check, I see that detekt default threshold for the class size is 600 lines (and it doesn't seem to be configured in our .detekt.yml), so I put the annotation as advised.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 8]
```diff
@@ -358,30 +366,34 @@ class FossId internal constructor(
         revision: String,
```
Comment (nnobelis) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:369]: Could you please add documentation to this function ?
Returning two values adds complexity and one cannot see straightaway what this function returns without the documentation.

File: scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt [hunk 9]
```diff
@@ -613,6 +621,7 @@ class FossIdTest : WordSpec({
             }
         }
 
+        @Suppress("LargeClass")
```
Comment (sschuberth) [inline scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt:624]: This annotation needs to go to the *class*, so before line 97.

Comment (yarosevych) [reply scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt]: Done

**Review by sschuberth — APPROVED** [review]
